File: linked_list/c/reverse_linked_list.c
#include "linked_list.h"

ListNode* reverse_linked_list(ListNode* head)
{
	ListNode *prev = 0, *curr = head, *next = 0;
	while(curr)
	{
		next = curr->next;
		curr->next = prev;
		prev = curr;
		curr = next;
	}
	return prev;
}

File: linked_list/py/reverse_linked_list.py
#      O -> O -> O -> $
# prev curr nxt
#                O -> $
#                prev curr
# Time:  O(n)
# Space: O(1)

def reverse_linked_list(head):
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev
